# Importing Libraries

In [87]:
import pandas as pd
import numpy as np
import random
from sklearn.metrics import accuracy_score   

# Data Pre-processing

In [88]:
dataset = pd.read_csv('Creditcard_data.csv')

In [89]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 772 entries, 0 to 771
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    772 non-null    int64  
 1   V1      772 non-null    float64
 2   V2      772 non-null    float64
 3   V3      772 non-null    float64
 4   V4      772 non-null    float64
 5   V5      772 non-null    float64
 6   V6      772 non-null    float64
 7   V7      772 non-null    float64
 8   V8      772 non-null    float64
 9   V9      772 non-null    float64
 10  V10     772 non-null    float64
 11  V11     772 non-null    float64
 12  V12     772 non-null    float64
 13  V13     772 non-null    float64
 14  V14     772 non-null    float64
 15  V15     772 non-null    float64
 16  V16     772 non-null    float64
 17  V17     772 non-null    float64
 18  V18     772 non-null    float64
 19  V19     772 non-null    float64
 20  V20     772 non-null    float64
 21  V21     772 non-null    float64
 22  V2

In [90]:
dataset.describe()

Time          V1          V2          V3          V4          V5  \
count  772.000000  772.000000  772.000000  772.000000  772.000000  772.000000   
mean   283.005181   -0.176963    0.217169    0.875172    0.285628   -0.005029   
std    171.834196    1.294724    1.173401    1.031878    1.258758    1.098143   
min      0.000000   -6.093248  -12.114213   -5.694973   -4.657545   -6.631951   
25%    126.500000   -0.896416   -0.174684    0.308677   -0.460058   -0.534567   
50%    282.000000   -0.382618    0.285843    0.905435    0.395919   -0.116612   
75%    432.000000    1.110739    0.885745    1.532969    1.117559    0.452818   
max    581.000000    1.586093    5.267376    3.772857    4.075817    7.672544   

               V6          V7          V8          V9  ...         V21  \
count  772.000000  772.000000  772.000000  772.000000  ...  772.000000   
mean     0.159081    0.123329   -0.057547   -0.030384  ...    0.004888   
std      1.225682    0.852075    0.830144    0.878183  ...    0.609335   
min     -3.498447   -4.925568   -7.494658   -2.770089  ...   -4.134608   
25%     -0.630717   -0.296289   -0.167880   -0.517068  ...   -0.213746   
50%     -0.109581    0.116329    0.034755   -0.082270  ...   -0.075802   
75%      0.482972    0.575390    0.252395    0.412261  ...    0.095149   
max      5.122103    4.808426    2.134599    5.459274  ...    5.273420   

              V22         V23         V24         V25         V26         V27  \
count  772.000000  772.000000  772.000000  772.000000  772.000000  772.000000   
mean    -0.096995   -0.040344   -0.002501    0.114337    0.022782    0.023353   
std      0.607228    0.358724    0.621507    0.429667    0.484227    0.300934   
min     -2.776923   -3.553381   -1.867208   -1.389079   -1.243924   -2.377933   
25%     -0.525289   -0.176915   -0.379766   -0.166227   -0.313631   -0.047868   
50%     -0.076551   -0.048353    0.091886    0.143723   -0.026414    0.023199   
75%      0.307438    0.070085    0.426339    0.425798    0.260408    0.112199   
max      1.574750    3.150413    1.215279    1.136720    3.087444    2.490503   

              V28       Amount       Class  
count  772.000000   772.000000  772.000000  
mean    -0.017045    68.668290    0.011658  
std      0.278332   197.838269    0.107411  
min     -2.735623     0.000000    0.000000  
25%     -0.033083     5.987500    0.000000  
50%      0.021034    16.665000    0.000000  
75%      0.087023    55.527500    0.000000  
max      1.575380  3828.040000    1.000000  

[8 rows x 31 columns]

In [91]:
dataset.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      1  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [92]:
dataset['Class'].value_counts()

0    763
1      9
Name: Class, dtype: int64

In [93]:
x = dataset.loc[:,dataset.columns!='Class']
y = dataset.loc[:,dataset.columns=='Class']

# Oversampling (Because dataset is imbalanced)

In [94]:
from imblearn.over_sampling import RandomOverSampler
rs = RandomOverSampler(random_state=50)
x_rs, y_rs = rs.fit_resample(x, y)

In [95]:
y_rs.value_counts()

Class
0        763
1        763
dtype: int64

In [96]:
x_rs['Class'] = y_rs

In [97]:
x_rs

Time        V1        V2        V3        V4        V5        V6  \
0        0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1        0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2        1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3        1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4        2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
...    ...       ...       ...       ...       ...       ...       ...   
1521   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
1522   484 -0.928088  0.398194  1.741131  0.182673  0.966387 -0.901004   
1523   164  0.073497  0.551033  0.451890  0.114964  0.822947  0.251480   
1524     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
1525     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   

            V7        V8        V9  ...       V21       V22       V23  \
0     0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474   
1    -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
2     0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412   
3     0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321   
4     0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458   
...        ...       ...       ...  ...       ...       ...       ...   
1521  0.086781 -0.202836  0.035154  ... -0.287592 -0.832682  0.128083   
1522  0.879016 -0.156590 -0.142117  ...  0.066353  0.281378 -0.257966   
1523  0.296319  0.139497 -0.123050  ... -0.128758 -0.381932  0.151012   
1524 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
1525 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   

           V24       V25       V26       V27       V28  Amount  Class  
0     0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1    -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      1  
2    -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3    -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4     0.141267 -0.206010  0.502292  0.219422  0.215153   69.99      0  
...        ...       ...       ...       ...       ...     ...    ...  
1521  0.339427  0.215944  0.094704 -0.023354  0.030892    2.69      1  
1522  0.385384  0.391117 -0.453853 -0.104448 -0.125765    1.00      1  
1523 -1.363967 -1.389079  0.075412  0.231750  0.230171    0.99      1  
1524 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      1  
1525 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      1  

[1526 rows x 31 columns]

ML MODELS ARE LOGISTIC REGRESSION ,DECISION TREE,NAIVE BYES,SUPPORT VECTOR MACHINE,KNN

In [98]:
#make a empty dataframe of columns having sampling method names and rows having model names and accuracy score
#column names are Simple Random Sampling,Straified Random Sampling,Systematic Sampling ,Cluster Sampling,Multistage Sampling
#row names are Logistic Regression,Decision Tree,Random Forest,Naive Bayes,Support Vector Machine
df = pd.DataFrame(index=['Logistic Regression','Decision Tree','Naive Bayes','Support Vector Machine',"KNN"],columns=['Simple Random Sampling','Stratified Random Sampling','Systematic Sampling','Cluster Sampling','Multi-Stage Sampling'])

# Sampling methods

**Sampling Method - 1 (Simple Random Sampling)**

In [99]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

# Simple Random Sampling
srs_sample = x_rs.sample(n = 1067, random_state=42)

# x, y division and train-test split
x = srs_sample.loc[:,srs_sample.columns!='Class']
y = srs_sample.loc[:,srs_sample.columns=='Class']
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=42)


In [100]:
# Logistic Regression
lr = LogisticRegression()
lr.fit(x_train, y_train)
y_pred = lr.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Simple Random Sampling']['Logistic Regression'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.8876404494382022


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\VS CODE\Python\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [101]:
# Decision Tree Classifier
dt = DecisionTreeClassifier()
dt.fit(x_train, y_train)
y_pred = dt.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Simple Random Sampling']['Decision Tree'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.9850187265917603


In [102]:
# Gaussian Naive Bayes
gnb = GaussianNB()
gnb.fit(x_train, y_train)
y_pred = gnb.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Simple Random Sampling']['Naive Bayes'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.7490636704119851


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [103]:
# Support Vector Machine
svm = SVC()
svm.fit(x_train, y_train)
y_pred = svm.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Simple Random Sampling']['Support Vector Machine'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.6891385767790262


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [104]:
# KNN
knn = KNeighborsClassifier()
knn.fit(x_train, y_train)
y_pred = knn.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Simple Random Sampling']['KNN'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.9812734082397003


d:\VS CODE\Python\lib\site-packages\sklearn\neighbors\_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [105]:
df

Simple Random Sampling Stratified Random Sampling  \
Logistic Regression                   0.88764                        NaN   
Decision Tree                        0.985019                        NaN   
Naive Bayes                          0.749064                        NaN   
Support Vector Machine               0.689139                        NaN   
KNN                                  0.981273                        NaN   

                       Systematic Sampling Cluster Sampling  \
Logistic Regression                    NaN              NaN   
Decision Tree                          NaN              NaN   
Naive Bayes                            NaN              NaN   
Support Vector Machine                 NaN              NaN   
KNN                                    NaN              NaN   

                       Multi-Stage Sampling  
Logistic Regression                     NaN  
Decision Tree                           NaN  
Naive Bayes                             NaN  
Support Vector Machine                  NaN  
KNN                                     NaN

**Sampling Method - 2 (Stratified Sampling)**

In [106]:
from sklearn import tree
from sklearn import linear_model
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

# Stratified Sampling
strat_sample = x_rs.groupby('Class', group_keys=False).apply(lambda x: x.sample(392,replace=True,random_state=42))

# x, y division and train-test split
x = strat_sample.loc[:,strat_sample.columns!='Class']
y = strat_sample.loc[:,strat_sample.columns=='Class']
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=42)


In [107]:
# Logistic Regression
lr = LogisticRegression()
lr.fit(x_train, y_train)
y_pred = lr.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Stratified Random Sampling']['Logistic Regression'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.9132653061224489


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\VS CODE\Python\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [108]:
# Decision Tree Classifier
dt = DecisionTreeClassifier()
dt.fit(x_train, y_train)
y_pred = dt.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Stratified Random Sampling']['Decision Tree'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.9693877551020408


In [109]:
#Gaussian Naive Bayes
gnb = GaussianNB()
gnb.fit(x_train, y_train)
y_pred = gnb.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Stratified Random Sampling']['Naive Bayes'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.7602040816326531


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [110]:
# Support Vector Machine
svm = SVC()
svm.fit(x_train, y_train)
y_pred = svm.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Stratified Random Sampling']['Support Vector Machine'] = accuracy_score(y_test, y_pred)

d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Accuracy Score:  0.6173469387755102


In [111]:
# KNN
knn = KNeighborsClassifier()
knn.fit(x_train, y_train)
y_pred = knn.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Stratified Random Sampling']['KNN'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.9591836734693877


d:\VS CODE\Python\lib\site-packages\sklearn\neighbors\_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [112]:
df

Simple Random Sampling Stratified Random Sampling  \
Logistic Regression                   0.88764                   0.913265   
Decision Tree                        0.985019                   0.969388   
Naive Bayes                          0.749064                   0.760204   
Support Vector Machine               0.689139                   0.617347   
KNN                                  0.981273                   0.959184   

                       Systematic Sampling Cluster Sampling  \
Logistic Regression                    NaN              NaN   
Decision Tree                          NaN              NaN   
Naive Bayes                            NaN              NaN   
Support Vector Machine                 NaN              NaN   
KNN                                    NaN              NaN   

                       Multi-Stage Sampling  
Logistic Regression                     NaN  
Decision Tree                           NaN  
Naive Bayes                             NaN  
Support Vector Machine                  NaN  
KNN                                     NaN

**Sampling Method - 3 (Systematic Sampling)**

In [113]:
from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn import tree
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB

# Systematic Sampling
sys_sample_df = x_rs.iloc[::5]

# x, y division and train-test split
x = sys_sample_df.loc[:,sys_sample_df.columns!='Class']
y = sys_sample_df.loc[:,sys_sample_df.columns=='Class']
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=42)


In [114]:
# Logistic Regression
lr = LogisticRegression()
lr.fit(x_train, y_train)
y_pred = lr.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Systematic Sampling']['Logistic Regression'] = accuracy_score(y_test, y_pred)

d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\VS CODE\Python\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Accuracy Score:  0.7922077922077922


In [115]:
# Decision Tree Classifier
dt = DecisionTreeClassifier()
dt.fit(x_train, y_train)
y_pred = dt.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Systematic Sampling']['Decision Tree'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.922077922077922


In [116]:
# Gaussian Naive Bayes
gnb = GaussianNB()
gnb.fit(x_train, y_train)
y_pred = gnb.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Systematic Sampling']['Naive Bayes'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.7792207792207793


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [117]:
# Support Vector Machine
svm = SVC()
svm.fit(x_train, y_train)
y_pred = svm.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Systematic Sampling']['Support Vector Machine'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.6103896103896104


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [118]:
# KNN
knn = KNeighborsClassifier()
knn.fit(x_train, y_train)
y_pred = knn.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Systematic Sampling']['KNN'] = accuracy_score(y_test, y_pred)


Accuracy Score:  0.922077922077922


d:\VS CODE\Python\lib\site-packages\sklearn\neighbors\_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [119]:
df

Simple Random Sampling Stratified Random Sampling  \
Logistic Regression                   0.88764                   0.913265   
Decision Tree                        0.985019                   0.969388   
Naive Bayes                          0.749064                   0.760204   
Support Vector Machine               0.689139                   0.617347   
KNN                                  0.981273                   0.959184   

                       Systematic Sampling Cluster Sampling  \
Logistic Regression               0.792208              NaN   
Decision Tree                     0.922078              NaN   
Naive Bayes                       0.779221              NaN   
Support Vector Machine             0.61039              NaN   
KNN                               0.922078              NaN   

                       Multi-Stage Sampling  
Logistic Regression                     NaN  
Decision Tree                           NaN  
Naive Bayes                             NaN  
Support Vector Machine                  NaN  
KNN                                     NaN

**Sampling Method - 4 (Cluster Sampling)**

In [120]:
# Cluster Sampling
def get_clustered_Sample(df, n_per_cluster, num_select_clusters):
    N = len(df)
    K = int(N/n_per_cluster)
    data = None
    for k in range(K):
        sample_k = df.sample(n_per_cluster)
        sample_k["cluster"] = np.repeat(k,len(sample_k))
        df = df.drop(index = sample_k.index)
        data = pd.concat([data,sample_k],axis = 0)
    random_chosen_clusters = np.random.randint(0,K,size = num_select_clusters)
    samples = data[data.cluster.isin(random_chosen_clusters)]
    return(samples)
clust_sample = get_clustered_Sample(df = x_rs, n_per_cluster = 20, num_select_clusters = 28)

# x, y division and train-test split
x = clust_sample.loc[:,clust_sample.columns!='Class']
y = clust_sample.loc[:,clust_sample.columns=='Class']
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=42)

In [121]:
# Logistic Regression
lr = LogisticRegression()
lr.fit(x_train, y_train)
y_pred = lr.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Cluster Sampling']['Logistic Regression'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.8956521739130435


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\VS CODE\Python\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [122]:
# Decision Tree Classifier
dt = DecisionTreeClassifier()
dt.fit(x_train, y_train)
y_pred = dt.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Cluster Sampling']['Decision Tree'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.9826086956521739


In [123]:
# Gaussian Naive Bayes
gnb = GaussianNB()
gnb.fit(x_train, y_train)
y_pred = gnb.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Cluster Sampling']['Naive Bayes'] = accuracy_score(y_test, y_pred)


Accuracy Score:  0.6521739130434783


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [124]:
# Support Vector Machine
svm = SVC()
svm.fit(x_train, y_train)
y_pred = svm.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Cluster Sampling']['Support Vector Machine'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.6695652173913044


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [125]:
# KNN
knn = KNeighborsClassifier()
knn.fit(x_train, y_train)
y_pred = knn.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Cluster Sampling']['KNN'] = accuracy_score(y_test, y_pred)

Accuracy Score:  0.8


d:\VS CODE\Python\lib\site-packages\sklearn\neighbors\_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [126]:
df

Simple Random Sampling Stratified Random Sampling  \
Logistic Regression                   0.88764                   0.913265   
Decision Tree                        0.985019                   0.969388   
Naive Bayes                          0.749064                   0.760204   
Support Vector Machine               0.689139                   0.617347   
KNN                                  0.981273                   0.959184   

                       Systematic Sampling Cluster Sampling  \
Logistic Regression               0.792208         0.895652   
Decision Tree                     0.922078         0.982609   
Naive Bayes                       0.779221         0.652174   
Support Vector Machine             0.61039         0.669565   
KNN                               0.922078              0.8   

                       Multi-Stage Sampling  
Logistic Regression                     NaN  
Decision Tree                           NaN  
Naive Bayes                             NaN  
Support Vector Machine                  NaN  
KNN                                     NaN

**Sampling Method - 5 Multi-Stage Sampling (Cluster and Simple Random)**

In [127]:
# Multi-Stage Sampling (Cluster sampling then simple random sampling)
def get_clustered_Sample(df, n_per_cluster, num_select_clusters):
    N = len(df)
    K = int(N/n_per_cluster)
    data = None
    for k in range(K):
        sample_k = df.sample(n_per_cluster)
        sample_k["cluster"] = np.repeat(k,len(sample_k))
        df = df.drop(index = sample_k.index)
        data = pd.concat([data,sample_k],axis = 0)
    random_chosen_clusters = np.random.randint(0,K,size = num_select_clusters)
    samples = data[data.cluster.isin(random_chosen_clusters)]
    return(samples)
clust_sample = get_clustered_Sample(df = x_rs, n_per_cluster = 20, num_select_clusters = 20)
mix_sample = clust_sample.sample(n = 200, random_state=42)

# x and y division
x = mix_sample.loc[:,mix_sample.columns!='Class']
y = mix_sample.loc[:,mix_sample.columns=='Class']

# Training-Testing Splitting
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=42)

# Logistic Regression
lr = LogisticRegression()
lr.fit(x_train, y_train)
y_pred = lr.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Multi-Stage Sampling']['Logistic Regression'] = accuracy_score(y_test, y_pred)

# Decision Tree Classifier
dt = DecisionTreeClassifier()
dt.fit(x_train, y_train)
y_pred = dt.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Multi-Stage Sampling']['Decision Tree'] = accuracy_score(y_test, y_pred)

# Gaussian Naive Bayes
gnb = GaussianNB()
gnb.fit(x_train, y_train)
y_pred = gnb.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Multi-Stage Sampling']['Naive Bayes'] = accuracy_score(y_test, y_pred)

# Support Vector Machine
svm = SVC()
svm.fit(x_train, y_train)
y_pred = svm.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Multi-Stage Sampling']['Support Vector Machine'] = accuracy_score(y_test, y_pred)

# KNN
knn = KNeighborsClassifier()
knn.fit(x_train, y_train)
y_pred = knn.predict(x_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
#add accuracy score to dataframe
df['Multi-Stage Sampling']['KNN'] = accuracy_score(y_test, y_pred)


Accuracy Score:  0.86
Accuracy Score:  0.94
Accuracy Score:  0.72
Accuracy Score:  0.6
Accuracy Score:  0.86


d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\VS CODE\Python\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
d:\VS CODE\Python\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = co

In [128]:
df

Simple Random Sampling Stratified Random Sampling  \
Logistic Regression                   0.88764                   0.913265   
Decision Tree                        0.985019                   0.969388   
Naive Bayes                          0.749064                   0.760204   
Support Vector Machine               0.689139                   0.617347   
KNN                                  0.981273                   0.959184   

                       Systematic Sampling Cluster Sampling  \
Logistic Regression               0.792208         0.895652   
Decision Tree                     0.922078         0.982609   
Naive Bayes                       0.779221         0.652174   
Support Vector Machine             0.61039         0.669565   
KNN                               0.922078              0.8   

                       Multi-Stage Sampling  
Logistic Regression                    0.86  
Decision Tree                          0.94  
Naive Bayes                            0.72  
Support Vector Machine                  0.6  
KNN                                    0.86

# Result Table

In [129]:
#index the datafram and name columns as Models 
df

Simple Random Sampling Stratified Random Sampling  \
Logistic Regression                   0.88764                   0.913265   
Decision Tree                        0.985019                   0.969388   
Naive Bayes                          0.749064                   0.760204   
Support Vector Machine               0.689139                   0.617347   
KNN                                  0.981273                   0.959184   

                       Systematic Sampling Cluster Sampling  \
Logistic Regression               0.792208         0.895652   
Decision Tree                     0.922078         0.982609   
Naive Bayes                       0.779221         0.652174   
Support Vector Machine             0.61039         0.669565   
KNN                               0.922078              0.8   

                       Multi-Stage Sampling  
Logistic Regression                    0.86  
Decision Tree                          0.94  
Naive Bayes                            0.72  
Support Vector Machine                  0.6  
KNN                                    0.86

# Conclusion
The Highest Accuracy achieved is DECISION TREE with SIMPLE RANDOM SAMPLING with 0.985019

In [133]:
#write a description of the entire notebook
description = """
This notebook is a comparison of different sampling techniques on the credit card fraud dataset.
The dataset is highly imbalanced and the techniques used are:
1. Simple Random Sampling
2. Stratified Sampling
3. Systematic Sampling
4. Cluster Sampling
5. Multi-Stage Sampling 

The models used are:
1. Logistic Regression
2. Decision Tree Classifier
3. Gaussian Naive Bayes
4. Support Vector Machine
5. KNN
"""
"""Write the dataframe to text"""
#also write this submission by Rishab Chakrabarti
description1 = """
By Understanding this we get highest accuracy for Decision Tree Classifier in Simple Random Sampling with 98.5019% accuracy.
This was created by Rishab Chakrabarti, Roll no:102003688"""
#write the description to a text file and the dataframe 
with open('submission.txt', 'w') as f:
    f.write(description)
    f.write(df.to_string())
    f.write(description1)
